les/serverpool.go

```diff
@@ -17,904 +17,283 @@
-			break
-		}
-		pool.dial((*poolEntry)(entry.(*knownEntry)), true)
-	}
-	for pool.knownSelected+pool.newSelected < targetServerCount {
-		entry := pool.newSelect.Choose()
-		if entry == nil {
-			break
-		}
-		pool.dial((*poolEntry)(entry.(*discoveredEntry)), false)
-	}
-	if fillWithKnownSelects {
-		// no more newly discovered nodes to select and since fast discover period
-		// is over, we probably won't find more in the near future so select more
-		// known entries if possible
-		for pool.knownSelected < targetServerCount {
-			entry := pool.knownSelect.Choose()
-			if entry == nil {
-				break
-			}
-			pool.dial((*poolEntry)(entry.(*knownEntry)), true)
-		}
-	}
+// registerPeer implements serverPeerSubscriber
+func (s *serverPool) registerPeer(p *serverPeer) {
+	s.ns.UpdateState(p.ID(), s.stConnected, s.stDialed, 0)
+	p.setValueTracker(s.vt, s.vt.Register(p.ID()))
+	p.updateVtParams()
 }
 
-// dial initiates a new connection
-func (pool *serverPool) dial(entry *poolEntry, knownSelected bool) {
-	if pool.server == nil || entry.state != psNotConnected {
-		return
-	}
-	entry.state = psDialed
-	entry.knownSelected = knownSelected
-	if knownSelected {
-		pool.knownSelected++
-	} else {
-		pool.newSelected++
+// unregisterPeer implements serverPeerSubscriber
+func (s *serverPool) unregisterPeer(p *serverPeer) {
+	if s.nodeWeight(p.ID(), true) >= nodeWeightThreshold {
+		s.ns.UpdateState(p.ID(), s.stHasValue, 0, 0)
 	}
-	addr := entry.addrSelect.Choose().(*poolEntryAddress)
-	log.Debug("Dialing new peer", "lesaddr", entry.node.ID().String()+"@"+addr.strKey(), "set", len(entry.addr), "known", knownSelected)
-	entry.dialed = addr
-	go func() {
-		pool.server.AddPeer(entry.node)
-		select {
-		case <-pool.closeCh:
-		case <-time.After(dialTimeout):
-			select {
-			case <-pool.closeCh:
-			case pool.timeout <- entry:
-			}
-		}
-	}()
+	s.ns.UpdateState(p.ID(), s.stRedialWait, s.stConnected, time.Second*10)
+	s.vt.Unregister(p.ID())
+	p.setValueTracker(nil, nil)
 }
 
-// checkDialTimeout checks if the node is still in dialed state and if so, resets it
-// and adjusts connection statistics accordingly.
-func (pool *serverPool) checkDialTimeout(entry *poolEntry) {
-	if entry.state != psDialed {
-		return
+// getTimeout calculates the current recommended timeout. This value is used by
+// the client as a "soft timeout" value. It also affects the service value calculation
+// of individual nodes.
+func (s *serverPool) getTimeout() time.Duration {
+	now := s.clock.Now()
+	s.timeoutLock.RLock()
+	timeout := s.timeout
+	refreshed := s.timeoutRefreshed
+	s.timeoutLock.RUnlock()
+	if refreshed != 0 && time.Duration(now-refreshed) < timeoutRefresh {
+		return timeout
 	}
-	log.Debug("Dial timeout", "lesaddr", entry.node.ID().String()+"@"+entry.dialed.strKey())
-	entry.state = psNotConnected
-	if entry.knownSelected {
-		pool.knownSelected--
-	} else {
-		pool.newSelected--
+	rts := s.vt.RtStats()
+	rts.Add(time.Second*2, 10, s.vt.StatsExpFactor())
```

Comment:
Any particular reason to add this "fake" statistic?

Reply:
It is an easy way to initialize with some conservative values when the database is new and we have no statistics. As soon as we have a considerable amount of real stats this small value won't matter.